Userland/Libraries/LibWeb/HTML/Parser/HTMLEncodingDetection.cpp

```diff
@@ -336,42 +336,80 @@ Optional<ByteString> run_prescan_byte_stream_algorithm(DOM::Document& document,
+    //    to obtain data to determine the encoding, then the cost of the delay could outweigh any performance improvements from the preparse.
+
+    // 4. If the transport layer specifies a character encoding, and it is supported, return that encoding with the confidence certain.
+    if (maybe_mime_type.has_value()) {
+        // FIXME: This is awkward because lecacy_extract_an_encoding can not fail
+        auto maybe_transport_encoding = Fetch::Infrastructure::legacy_extract_an_encoding(maybe_mime_type, "invalid"sv);
+        if (maybe_transport_encoding != "invalid"sv)
+            return maybe_transport_encoding;
     }
 
-    // FIXME: If the HTML parser for which this algorithm is being run is associated with a Document whose browsing context
-    //        is non-null and a child browsing context.
-    // FIXME: If the user agent has information on the likely encoding for this page, e.g. based on the encoding of the page
-    //        when it was last visited.
-
+    // 5. Optionally prescan the byte stream to determine its encoding, with the end condition being when the user agent decides that scanning further bytes would not
+    //    be efficient. User agents are encouraged to only prescan the first 1024 bytes. User agents may decide that scanning any bytes is not efficient, in which case
+    //    these substeps are entirely skipped.
+    //    The aforementioned algorithm returns either a character encoding or failure. If it returns a character encoding, then return the same encoding, with confidence tentative.
+    auto prescan = run_prescan_byte_stream_algorithm(document, input);
+    if (prescan.has_value())
+        return prescan.value();
+
+    // 6. FIXME: If the HTML parser for which this algorithm is being run is associated with a Document d whose container document is non-null, then:
+    // 1. Let parentDocument be d's container document.
+    // 2. If parentDocument's origin is same origin with d's origin and parentDocument's character encoding is not UTF-16BE/LE, then return parentDocument's character
+    //    encoding, with the confidence tentative.
+
+    // 7. Otherwise, if the user agent has information on the likely encoding for this page, e.g. based on the encoding of the page when it was last visited, then return
+    //    that encoding, with the confidence tentative.
+
+    // 8. FIXME: The user agent may attempt to autodetect the character encoding from applying frequency analysis or other algorithms to the data stream. Such algorithms
+    //    may use information about the resource other than the resource's contents, including the address of the resource. If autodetection succeeds in determining a
+    //    character encoding, and that encoding is a supported encoding, then return that encoding, with the confidence tentative. [UNIVCHARDET]
     if (!Utf8View(StringView(input)).validate()) {
```

Comment:
I'm not sure if the intermediate StringView is needed but I left it there because I was also unsure if it really is unnecessary.